cpmpy/solvers/TEMPLATE.py

```diff
@@ -0,0 +1,283 @@
+#!/usr/bin/env python
+"""
+    Template file for a new solver interface
+
+    Replace <TEMPLATE> by the solver's name, and implement the missing pieces
+    The functions are ordered in a way that could be convenient to 
+    start from the top and continue in that order
+
+    TODO: smth about testing as you progress...
+    TODO: I guess we should add a __main__ that runs some tests?
+"""
+"""
+    Interface to TEMPLATE's API
+
+    <some information on the solver>
+
+    Documentation of the solver's own Python API:
+    <URL to docs or source code>
+
+    ===============
+    List of classes
+    ===============
+
+    .. autosummary::
+        :nosignatures:
+
+        CPM_template
+"""
+from .solver_interface import SolverInterface, SolverStatus, ExitStatus
+from ..expressions.core import *
+from ..expressions.variables import _BoolVarImpl, NegBoolView
+from ..expressions.utils import is_any_list
+from ..transformations.get_variables import get_variables
+
+class CPM_template(SolverInterface):
+    """
+    Interface to TEMPLATE's API
+
+    Requires that the 'TEMPLATEpy' python package is installed:
+    $ pip install TEMPLATEpy
+
+    See detailed installation instructions at:
+    <URL to detailed solver installation instructions, if any>
+
+    Creates the following attributes:
+    user_vars: set(), variables in the original (non-transformed) model,
+                    for reverse mapping the values after `solve()`
+    cpm_status: SolverStatus(), the CPMpy status after a `solve()`
+    tpl_model: object, TEMPLATE's model object
+    _varmap: dict(), maps cpmpy variables to native solver variables
+    """
+
+    @staticmethod
+    def supported():
+        # try to import the package
+        try:
+            import TEMPLATEpy as gp
+            return True
+        except ImportError as e:
+            return False
+
+
+    def __init__(self, cpm_model=None, solver=None):
+        """
+        Constructor of the native solver object
+
+        Arguments:
+        - cpm_model: a CPMpy Model()
+        """
+        if not self.supported():
+            raise Exception("CPM_TEMPLATE: Install the python package 'TEMPLATEpy'")
+
+        self.name = "TEMPLATE" # solver's name
+
+        # initialise the native solver object
+        self.tpl_model = TEMPLATEpy.Model("cpmpy")
+
+        # initialise everything else and post the constraints/objective
+        # it is sufficient to implement __add__() and minimize/maximize() below
+        super().__init__(cpm_model, solver)
+
+        # TODO: these 4 in super()
+        self.cpm_status = SolverStatus(self.name)
+
+        # initialise variable handling
+        self.user_vars = set() # variables in the original (non-transformed) model
```

Comment:
Right now, user_vars is a list. Altough I see no reason why it couldn't be a set?

Reply:
I think in MiniZinc it was a list so that we could output the variables 'in order' of use.

For all other solvers it should be a set, as that makes updating the sets on incremental addition more efficient. So, do switch to set and we'll see in the MiniZinc interface if that breaks something.